lib/obj.js

```diff
@@ -1,365 +1,479 @@
+            return;
+        }
+        return loadMtl(mtlPath)
+            .then(function(materialsInMtl) {
+                materials = combine(materials, materialsInMtl);
+            })
+            .catch(function() {
+                options.logger('Could not read mtl file at ' + mtlPath + '. Using default material instead.');
             });
-        });
+    }).then(function() {
+        return materials;
     });
 }
 
-function getImages(inputPath, materials) {
-    // Collect all the image files from the materials
-    var images = [];
+function loadImages(imagePaths, objPath, options) {
+    var images = {};
+    return Promise.map(imagePaths, function(imagePath) {
+        if (options.secure && outsideDirectory(imagePath, objPath)) {
+            options.logger('Could not read image file at ' + imagePath + ' because it is outside of the obj directory and the secure flag is true. Material will ignore this image.');
+            return;
+        }
+        return loadImage(imagePath, options)
+            .then(function(image) {
+                if (defined(image)) {
+                    images[imagePath] = image;
+                }
+            })
+            .catch(function() {
+                options.logger('Could not read image file at ' + imagePath + '. Material will ignore this image.');
+                return undefined;
+            });
+    }).then(function() {
+        return images;
+    });
+}
+
+function getImagePaths(materials) {
+    var imagePaths = {};
     for (var name in materials) {
         if (materials.hasOwnProperty(name)) {
             var material = materials[name];
-            if (defined(material.ambientColorMap) && (images.indexOf(material.ambientColorMap) === -1)) {
-                images.push(material.ambientColorMap);
+            if (defined(material.ambientColorMap)) {
+                imagePaths[material.ambientColorMap] = true;
             }
-            if (defined(material.diffuseColorMap) && (images.indexOf(material.diffuseColorMap) === -1)) {
-                images.push(material.diffuseColorMap);
+            if (defined(material.diffuseColorMap)) {
+                imagePaths[material.diffuseColorMap] = true;
             }
-            if (defined(material.emissionColorMap) && (images.indexOf(material.emissionColorMap) === -1)) {
-                images.push(material.emissionColorMap);
+            if (defined(material.emissionColorMap)) {
+                imagePaths[material.emissionColorMap] = true;
             }
-            if (defined(material.specularColorMap) && (images.indexOf(material.specularColorMap) === -1)) {
-                images.push(material.specularColorMap);
+            if (defined(material.specularColorMap)) {
+                imagePaths[material.specularColorMap] = true;
             }
         }
     }
+    return Object.keys(imagePaths);
```

Comment:
In the case of different material texture slots referencing the same textures it was easier to handle duplicates with a map rather than doing `indexOf` checks.